### ISSUE TITLE
Maps between local and global vertex indices in a ParMesh?

### ISSUE THREAD

**Issue by kennyweiss** [mfem/mfem#1102]
PR #1058 adds maps between local and global element indices. (Thanks @rw-anderson!)

Does the corresponding functionality exist somewhere for vertices? 
E.g. in the ``NCMesh`` classes, in setup code for load balancing/partitioning or in a miniapp/user application? (Or perhaps even in ``ParMesh`` and I'm somehow missing it?)

I am trying to use a ``ParMesh`` with a library that rebalances a subset of the mesh and its API requires global indices for the elements and vertices on each domain.

**Comment by kennyweiss** [comment]
Following a thread from @jakubcerveny's comment here https://github.com/mfem/mfem/issues/939#issuecomment-497458438

* Is ``ParFiniteElementSpace``'s TrueDofs what I'm looking for?
* Any chance of adding equivalent functionality to ``ParMesh``?

**Comment by jakubcerveny** [comment]
I think the only reasonable way for the user to get global vertex numbering is to create a linear H1 space, query an element's local DOFs and convert them to global DOFs with `ParFiniteElementSpace::GetGlobalTDofNumber`.

It's unlikely this will be added to `ParMesh`, since globally enumerating vertices, edges and faces is basically the main job of `ParFiniteElementSpace`.

**Comment by jakubcerveny** [comment]
I wouldn't be afraid of constructing the `ParFiniteElementSpace`, it's a relatively lightweight structure on top of `ParMesh`...

**Comment by kennyweiss** [comment]
Thanks @jakubcerveny that seems straightforward enough.

**Comment by kennyweiss** [comment]
Closing for now... will reopen if necessary.

**Comment by kennyweiss** [comment]
@jakubcerveny -- I was able to get a solution with ``ParFiniteElementSpace::GetGlobalTDofNumber`` for conforming meshes, but ran into a problem with non-conforming meshes.

I originally tried simply looping through the dofs and getting the true dofs:

```cpp
     // Get global vertex IDs
     {
       auto* fec = new mfem::H1_FECollection(1, dim);
       auto* fes = new mfem::ParFiniteElementSpace(mesh, fec);

       std::cout << "Printing vertex tdofs:";
       for(int i=0; i< mesh->GetNV(); ++i)
       {
          auto tdof = fes->GetGlobalTDofNumber(i);  // PROBLEM HERE!

           std::cout<<"\t Vert: " << i
                    << "; global dof " << tdof 
                    << std::endl;
       }
       EXO_INFO("---");

       delete fes;
       delete fec;
     }
```
And got the following error:
```bash
Verification failed: (ldof_ltdof[ldof] >= 0) is false:
 --> ldof 0 not a true DOF. ( Value: -1 )
 ... in function: int mfem::ParFiniteElementSpace::GetGlobalTDofNumber(int) const
 ... in file: mfem-5edd742/fem/pfespace.cpp:789
```

Then, I thought maybe it would work if I loop the the local dofs indices of each element, but got the same error.
```cpp
     // Get global vertex IDs
     {
       auto* fec = new mfem::H1_FECollection(1, dim);
       auto* fes = new mfem::ParFiniteElementSpace(mesh, fec);

       std::cout << "Printing vertex tdofs:";
       Array<int> idofs;

       for(int i=0; i< mesh->GetNE(); ++i)
       {
          fes->GetElementDofs(i, idofs);
          for(int d=0; d< idofs.Size(); ++d)
          {
             auto ldof = idofs[d];
             auto tdof = fes->GetGlobalTDofNumber(ldof);  // PROBLEM HERE!
             
             std::cout<<"\t El: " << i
                      <<"; local dof: " << ldof
                      <<"; global dof " << tdof 
                      << std::endl;
          }
       }

       delete fes;
       delete fec;
     }
```

Here is the relevant section of ``pfespace.cpp``:
https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/pfespace.cpp#L786-L807

**Comment by kennyweiss** [comment]
In case it's helpful, here is a description of my simplified problem setup and a screenshot.

I start with a mesh containing two linear quad elements that are in separate connected components, and then, after converting to an ``NCMesh``, I apply two levels of uniform refinement to the left block and only one to the right block.  

The underlying code calls (among other things):
```cpp
// Ensure non-conforming mesh (quads/hexes)
mesh->EnsureNCMesh();
...
Array<Refinement> refs;
...
mesh->GeneralRefinement(refs);
```

Here is a screenshot from a four rank run of the problem (with 16 elements in the left component of the mesh and 4 elements in the mesh's right component):
![ncmesh_tdof_problem](https://user-images.githubusercontent.com/5626552/67166065-fecfd280-f340-11e9-8e40-6cd99e69841f.png)

**Comment by v-dobrev** [comment]
In a non-conforming mesh some vertices are dependent so they do not have a corresponding global true-dof. You should be able to filter them by checking if `fes->GetLocalTDofNumber(i) >= 0` where `i` is a local vertex index, as in your first code snippet.

I'm not sure what the expected input for the rebalancing library is -- does it actually need to have global indices for these dependent vertices? And more generally, does it actually support non-conforming meshes? If it is using the vertex indices to determine a global element-to-element connectivity graph (which it then partitions), then the case of hanging nodes has to be handled in some way and it has to know which nodes are hanging and which their parent nodes/vertices are.

**Comment by rw-anderson** [comment]
Hi Kenny, I'm not sure exactly what the motivation was for this, but if it was to try to repartition nc meshes that end up with bad partitions after uniform refinement, these links may be helpful:

https://github.com/mfem/mfem/issues/1110
https://github.com/mfem/mfem/tree/uniform-refinement-element-ordering-dev

**Comment by kennyweiss** [comment]
Thanks for the suggestions @jakubcerveny, @v-dobrev and @rw-anderson.

Perhaps the rebalancing application was a red herring. What I am really after is unique global IDs for the distinct vertices, elements and boundary elements of a ``ParMesh``.  I need to be able to generate these for conforming and non-conforming meshes. They do not actually need to be related to the tdofs, in case that is helpful.

I coded up a parallel unit test as I was working through this issue -- see PR #1142.
I was able to get it to work for conforming meshes, but not ``NCMesh``es. Any help on this would be greatly appreciated.
